client/tests/integration/restart_peer.rs

```diff
@@ -55,6 +55,7 @@ fn restarted_peer_should_have_the_same_asset_amount() -> Result<()> {
             .find(|asset| asset.id().definition_id == asset_definition_id)
             .expect("Asset not found");
         assert_eq!(AssetValue::Quantity(quantity), *asset.value());
+        peer.stop();
```

Comment:
This design confuses me a bit.
After stopping, the peer goes into an undefined state, and all methods are still available to use. This may lead to misusing the peer or to even bugs imho.
It made me think about introducing StoppedPeer.
```rust
// drop the `peer` here
let stopped_peer: StoppedPeer = peer.stop();
// ....
// restart a new `peer` and drop the `stopped_peer`
let peer = stopped_peer.restart(...);
// or
let peer = PeerBuilder::new().restart(stopped_peer);
```

Reply:
Or maybe we can recreate peer in each scope? And rely on `Drop`.

Reply:
> the peer goes into an undefined state

It's defined I think - basically Iroha instance is dropped, and all the context for it isn't. You shouldn't get any problems reusing it. I'll have a look again to double-check if there's any strange behaviour one can trigger that way.

>  It made me think about introducing StoppedPeer.

That sound good though, having type indicate state instead of struct of `Options` is almost always a cleaner solution IMO. That said, I think it's out of scope for this if `stop` indeed doesn't lead to any problems.

Reply:
@Erigara @pesterev I looked through it once more and I'm convinced `stop` doesn't lead to any problematic behavior.

I agree it's not the best design, but I think we need to refactor `Peer` and `PeerBuilder` wholesale to do measurably better. What do you think?

Reply:
Got it. Let's go ahead with this.

Reply:
Agree with you let's return to this later.